http/request.lua

```diff
@@ -167,6 +167,146 @@ function request_methods:to_uri(with_userinfo)
 	return scheme .. "://" .. authority .. path
 end
 
+function request_methods:to_curl()
+	local cmd = {
+		"curl";
+	}
+	local n = 1
+
+	if self.version then
+		if self.version == 1 then
+			cmd[n+1] = "-0"
+		elseif self.version == 1.1 then
+			cmd[n+1] = "--http1.1"
+		elseif self.version == 2 then
+			cmd[n+1] = "--http2"
+		else
+			error("invalid version")
+		end
+		n = n + 1
+	end
+
+	if self.proxy then
+		if type(self.proxy) ~= "string" then
+			error("NYI")
+		end
+		cmd[n+1] = "--proxy"
+		cmd[n+2] = self.proxy
+		n = n + 2
+	elseif not self.proxies then
+		cmd[n+1] = "--noproxy"
+		cmd[n+2] = "*"
+		n = n + 2
+	elseif self.proxies ~= default_proxies then
+		assert(getmetatable(self.proxies) == http_proxies.mt, "proxies property should be a http.proxies object")
+		error("NYI")
+	end
+
+	if self.expect_100_timeout ~= 1 then
+		cmd[n+1] = "--expect100-timeout"
+		cmd[n+2] = string.format("%d", self.expect_100_timeout)
+		n = n + 2
+	end
+
+	if self.follow_redirects then
+		cmd[n+1] = "--location-trusted"
+		cmd[n+2] = "-e"
+		cmd[n+3] = ";auto"
+		n = n + 3
+
+		if self.max_redirects ~= 50 then -- curl default is 50
+			cmd[n+1] = "--max-redirs"
+			cmd[n+2] = string.format("%d", self.max_redirects or -1)
+			n = n + 2
+		end
+
+		if self.post301 then
+			cmd[n+1] = "--post301"
+			n = n + 1
+		end
+
+		if self.post302 then
+			cmd[n+1] = "--post302"
+			n = n + 1
+		end
+	end
+
+	if self.tls and self.tls ~= true then
+		error("NYI")
+	end
+
+	local scheme = self.headers:get(":scheme")
+	-- Unlike the ':to_uri' method, curl needs the authority in the URI to be the actual host/port
+	local authority = http_util.to_authority(self.host, self.port, scheme)
+	local path = self.headers:get(":path")
+	assert(path == "" or path:sub(1,1) == "/" or path:sub(1,1) == "?", "invalid path for cURL")
+	local url = scheme .. "://" .. authority .. path
+	if url:match("[%[%]%{%}]") then
+		-- Turn off curl URL globbing
+		cmd[n+1] = "-g"
+		n = n + 1
+	end
+	cmd[n+1] = url
+	n = n + 1
+
+	for name, value in self.headers:each() do
+		if name:sub(1,1) == ":" then
+			if name == ":authority" then
+				if value ~= authority then
+					cmd[n+1] = "-H"
+					cmd[n+2] = "host: " .. value
+					n = n + 2
+				end
+			elseif name == ":method" then
+				if value == "HEAD" then
+					cmd[n+1] = "-I"
+					n = n + 1
+				elseif (value ~= "GET" or self.body ~= nil) and (value ~= "POST" or self.body == nil) then
+					cmd[n+1] = "-X"
+					cmd[n+2] = value
+					n = n + 2
+				end
+			end
+		elseif name == "user-agent" then
+			cmd[n+1] = "-A"
```

Comment:
If user agent not set then need to tell curl not to send one. `-H "User-Agent;"`